pythonFiles/include/blender_vscode/installation.py

```diff
@@ -1,55 +1,93 @@
+import importlib.metadata
+import shutil
 import sys
 import subprocess
-
+import textwrap
 import bpy
 
 from pathlib import Path
+from typing import List, Optional, Tuple
 
-from . import handle_fatal_error
 from .environment import python_path
+from semver import Version
 
 _CWD_FOR_SUBPROCESSES = python_path.parent
 
 
-def ensure_packages_are_installed(package_names):
-    if packages_are_installed(package_names):
-        return
-
-    install_packages(package_names)
+def assert_packages_are_installed(package_names: List[str]):
+    from . import handle_fatal_error
 
+    for package in package_names:
+        if not is_module_installed(package):
+            handle_fatal_error(
+                textwrap.dedent(
+                    f"""\
+            Package can not be imported: {package}
+            If you are experiencing PermissionError please scroll up to the line `Execute:` and execute the command manually.
+            Also close all Blender instances.
+            For old system wide Blender installations you might to reinstall Blender."""
+                )
+            )
 
-def packages_are_installed(package_names):
-    return all(module_can_be_imported(name) for name in package_names)
 
-
-def install_packages(package_names):
-    if not module_can_be_imported("pip"):
+def ensure_packages_are_installed(package_names: List[str]):
+    if not is_module_installed("pip"):
         install_pip()
 
-    for name in package_names:
-        ensure_package_is_installed(name)
-
-    assert packages_are_installed(package_names)
-
-
-def ensure_package_is_installed(name: str):
-    if not module_can_be_imported(name):
-        install_package(name)
-
+    if not any(is_module_installed(name) for name in package_names):
+        print(f"INFO: installing dependencies: {package_names}")
+        install_packages(package_names)
+        return
 
-def install_package(name: str):
+    # Perform precise checks if any dependency changes are needed.
+    # Manual checks prevent from unnecessary updates which on <0.21 installations may require admin rights
+    # Moreover upgrading when using --target is not possible.
+    requires_reinstall = False
+    for name in package_names:
+        _name, requested_version = _split_package_version(name)
+        if requested_version is None:
+            continue
+        requested_version = Version.parse(requested_version)
+        real_version = Version.parse(package_version(name))
+        assert isinstance(requested_version, Version), requested_version
+        assert isinstance(real_version, Version), real_version
+        if "==" in name and not (real_version == requested_version):
```

Comment:
Is there no builtin library that helps with these kinds of checks? It feels a bit weird to have to implement it manually.